File: 0429-n-ary-tree-level-order-traversal/breadth-first-search/cpp/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<vector<int>> levelOrder(Node* root) {
        if (!root) {
            return {};
        }

        vector<vector<int>> result;
        queue<Node*> q;
        q.push(root);

        while (!q.empty()) {
            int level_size = q.size();
            vector<int> level_res;

            for (int i = 0; i < level_size; i++) {
                Node* node = q.front();
                q.pop();
                level_res.push_back(node->val);

                for (Node* child : node->children) {
                    if (child) q.push(child);
                }
            }
            result.push_back(level_res);
        }
        return result;
    }
};

File: 0429-n-ary-tree-level-order-traversal/breadth-first-search/java/solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<Node> queue = new LinkedList<>();
        queue.add(root);
        
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> levelRes = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                Node node = queue.poll();
                levelRes.add(node.val);

                if (node.children != null) {
                    for (Node child : node.children) {
                        if (child != null) queue.add(child);
                    }
                }
            }
            result.add(levelRes);
        }
        return result;
    }
}
